state/transaction.go

```diff
@@ -1022,29 +1019,16 @@ func (s *State) EstimateGas(transaction *types.Transaction, senderAddress common
 		previousBatch = lastBatches[1]
 	}
 
-	lowEnd, err = core.IntrinsicGas(transaction.Data(), transaction.AccessList(), s.isContractCreation(transaction), true, false, false)
-	if err != nil {
-		return 0, nil, err
-	}
+	highEnd := s.cfg.MaxCumulativeGasUsed
 
-	if lowEnd == ethTransferGas && transaction.To() != nil {
-		code, err := s.tree.GetCode(ctx, *transaction.To(), stateRoot.Bytes())
-		if err != nil {
-			log.Warnf("error while getting transaction.to() code %v", err)
-		} else if len(code) == 0 {
-			return lowEnd, nil, nil
-		}
-	}
+	isZeroAddress := senderAddress.String() == ZeroAddress.String()
+	isDefaultSenderAddress := senderAddress.String() == common.HexToAddress(DefaultSenderAddress).String()
+	isSenderSpecified := !isZeroAddress && !isDefaultSenderAddress
+	isGasPriceSet := transaction.GasPrice().BitLen() != 0
 
-	if transaction.Gas() != 0 && transaction.Gas() > lowEnd {
-		highEnd = transaction.Gas()
-	} else {
-		highEnd = s.cfg.MaxCumulativeGasUsed
-	}
-
-	var availableBalance *big.Int
-
-	if senderAddress != ZeroAddress {
+	// if sender is specified and gas price is set
+	// try to set the high end to the max value the account balance can afford
+	if isSenderSpecified && isGasPriceSet {
 		senderBalance, err := s.tree.GetBalance(ctx, senderAddress, stateRoot.Bytes())
```

Comment:
This check should be done even if `from` is not specified. It will be done with the `defaultAddress`